### PR TITLE
Improvements to automatic tests

### PR DESCRIPTION
**Pull request by benlk**
## Changes

`.travis.yml`:
- copies formatting of https://github.com/npr/nprapi-wordpress/blob/master/.travis.yml
- cuts PHP versions down to the four versions that are not yet end-of-life: 5.6, 7.0, 7.1, ~~7.2~~ https://secure.php.net/supported-versions.php
    - previous versions of PHP contain unpatched security flaws
    - 5.6 will stop receiving security updates on 31 December 2018
    - 7.0 will stop receiving security updates on 3 December 2018
- switches WordPress versions to the latest patch release in each release's branch, and extends WordPress versions covered from 4.1 to the current release, 4.9

tests:
- Replaces http://publicmediaplatform.org/wp-content/uploads/logo1.png with http://blog.apps.npr.org/img/nprlogo.gif because the PMP logo image was 301'ing to a non-image URL.
- Adds a separate test for the output of `pmp_client_secret_input` when no PMP options have been saved
- Adds an explanation of why the test `test_pmp_verify_settings` might fail: if your test suite's env vars aren't set in accordance with bootstrap.php
- Patches `test_pmp_client_secret_input_option` so that it will run as intended even if your test suite's env vars aren't set in accordance with bootstrap.php

`pmp_client_secret()`:

- Flips the logic so that the default case is to output the `pmp_client_secret` field, and only output the reset button if the client secret has been saved

## Travis tests update:

This PR expands the versions tested because we want to make sure that the plugin is compatible with recent PHP and WordPress versions. The simple way to do that is to enumerate a list of WordPress versions and a list of PHP versions, and then run the tests at each intersection. Not only is this approach inefficient, it causes problems when old versions of WordPress include a test suite that assumes an old version of PHPUnit, which is not available when Travis CI only offers one version of PHPUnit per PHP version.

Based on research in npr/pmp-php-sdk#36,
Travis CI runs tests using the following PHPUnit versions:

- PHP 5.6: PHPUnit 5.7.23
- PHP 7: PHPUnit 6.4.3
- PHP 7.1: PHPUnit 6.4.3
- PHP 7.2: PHPUnit 7.0.2

Reasons that given versions of WordPress have tests incompatible with travis-ci.org:

1. There are PHP warnings about deprecated PHP-4-style constructors.
2. Requires PHPUnit 5 because PHPUnit 6 support was not included until WordPress 4.7.
3. Unless there's been an update since [this trac thread](https://core.trac.wordpress.org/ticket/39822), WordPress tests still assume PHPUnit 6, and Travis runs PHP 7.2 tests using PHPUnit 7

| phpunit | 5.7 | 6.4 | 6.4 | 7.0 |                 |                  |
|---------|-----|-----|-----|-----|-----------------|------------------|
| **php** | 5.6 | 7.0 | 7.1 | 7.2 |                 |                  |
| **WP**      |     |     |     |     | wp release year | then-current php |
| 4.1     |     | 1   | 1   | 1   | 2014            | 5.5              |
| 4.2     |     | 2   | 2   | 2   | 2015            | 5.6              |
| 4.3     |     | 2   | 2   | 2   | 2015            | 5.6              |
| 4.4     |     | 2   | 2   | 2   | 2015            | 7.0              |
| 4.5     |     | 2   | 2   | 2   | 2016            | 7.0              |
| 4.6     |     | 2   | 2   | 2   | 2016            | 7.0              |
| 4.7     |     | 2   | 2   | 2   | 2016            | 7.1              |
| 4.8     |     |     |     | 3   | 2017            | 7.1              |
| 4.9     |     |     |     | 3   | 2017            | 7.2              |

## Concerns

~~We're going from `( 4 PHP * 1 WordPress ) = 4 builds` to `( 4 PHP * 9 WordPress ) = 36 builds`. builds will take forever to run.~~ After ce1b173, the number of builds is limited to 13.

NPR needs to configure Travis builds as described in [the top comment](https://github.com/npr/pmp-wordpress-plugin/pull/131#issuecomment-373535530) on this pull request in order for tests to pass.

I don't know where the Slack notifications line in `.travis.yml` runs to.

### CODE REVIEW

**Comment by benlk** [comment]
```
1) TestFunctions::test_pmp_verify_settings
Either pmp_verify_settings() is broken, or the option value pmp_settings is not properly set by tests/bootstrap.php. Does your test-runner have the appropriate environment variables configured?
Failed asserting that false is true.
/tmp/wordpress/src/wp-content/plugins/pmp-wordpress-plugin/tests/inc/test-functions.php:106
```

Tests are probably failing because the following environment variables are not set at https://travis-ci.org/npr/pmp-wordpress-plugin/settings

```php
	$pmp_creds = array(
		'pmp_api_url' => getenv('PMP_API_URL'),
		'pmp_client_id' => getenv('PMP_CLIENT_ID'),
		'pmp_client_secret' => getenv('PMP_CLIENT_SECRET')
	);
```

I believe the valid values for `PMP_API_URL` are listed in inc/settings.php:

- https://api.pmp.io
- https://api-sandbox.pmp.io

https://github.com/npr/pmp-wordpress-plugin/blob/c710914cf8fccb49bc55a196bc62a0dfd3a98e17/inc/settings.php#L54-L55

**Comment by benlk** [comment]
Tests are presently broken in PHP 7 and above, because the PMP SDK vendors in a version of Symfony whose tests depend upon a class name that was removed from PHPUnit 6 and following. For more information about this problem, see https://github.com/npr/pmp-php-sdk/issues/36

**Comment by benlk** [comment]
Testing with WordPress 4.1, 4.2 and PHP versions `>= 7` results in a lot of noise in the testing log because PHP 7 [deprecated PHP-4-style class constructors](https://wiki.php.net/rfc/remove_php4_constructors), and WordPress didn't remove all of them until [WordPress 4.3](https://make.wordpress.org/core/2015/07/02/deprecating-php4-style-constructors-in-wordpress-4-3/) on August 18, 2015.

**Comment by benlk** [comment]
ce1b173 removes that noise and a number of guaranteed-to-fail builds, with the result that tests are now only failing because the env vars aren't configured. After that commit, I updated the PR description to explain why we're not running PHP >= 7.0 tests against WordPress < 4.8.

**Comment by benlk** [comment]
External pull requests will generally always fail, as described in #136. #136 was a clone of this pull request, but coming from within the `npr/pmp-wordpress-plugin` repository instead of being an external pull request from a fork of the repository. Because of https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions, the secret environment variables (the API client ID and API client secret) are not available to outside pull requests, and as a result outside pull requests will never be able to completely run the test suite.

This is problematic, but I've been working towards a better explanation of why those tests are failing, in the form of the `test_pmp_verify_settings_more()` test, which does *not* output those secrets, but does describe their length and `empty()` and `isset()` statuses, where `"1"` is true and `""` is false. This can be seen in the output of https://travis-ci.org/npr/pmp-wordpress-plugin/jobs/370227076:

```
1) TestFunctions::test_pmp_verify_settings
Either pmp_verify_settings() is broken, or the option value pmp_settings is not properly set by tests/bootstrap.php. Does your test-runner have the appropriate environment variables configured?
Failed asserting that false is true.
/tmp/wordpress/src/wp-content/plugins/pmp-wordpress-plugin/tests/inc/test-functions.php:106
2) TestFunctions::test_pmp_verify_settings_more
This is debug information for test_pmp_verify_settings: array (
  'pmp_api_url' => 'pmp_api_url: "26" long, empty "", isset "1"',
  'pmp_client_id' => 'pmp_client_id: "0" long, empty "1", isset "1"',
  'pmp_client_secret' => 'pmp_client_secret: "0" long, empty "1", isset "1"',
```

What this is saying is that the pull request does not have access to the client ID or secret.

Since external pull requests won't have access to those variables, external tests will likely fail. It's up to the plugin maintainers whether they want to copy outside branches into this repository and re-open the pull requests. Doing so will probably not be necessary in many cases.

In conclusion: I'm closing this pull request in favor of #136, so that tests will have a better chance at running successfully.
